File: app/src/main/java/at/bitfire/icsdroid/ui/AddCalendarValidationFragment.java
/*
 * Copyright (c) 2013 – 2016 Ricki Hirner (bitfire web engineering).
 *
 * This program is free software: you can redistribute it and/or modify it under the terms of the GNU
 * General Public License as published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
 * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE.  See the GNU General Public License for more details.
 *
 */

package at.bitfire.icsdroid.ui;

import android.app.Dialog;
import android.app.ProgressDialog;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.LoaderManager;
import android.support.v4.content.AsyncTaskLoader;
import android.support.v4.content.Loader;
import android.util.Base64;
import android.util.Log;
import android.widget.Toast;

import org.apache.commons.lang3.StringUtils;

import java.io.IOException;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.net.URLConnection;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import javax.net.ssl.HttpsURLConnection;

import at.bitfire.ical4android.Event;
import at.bitfire.ical4android.InvalidCalendarException;
import at.bitfire.icsdroid.Constants;
import at.bitfire.icsdroid.MTMLoader;
import at.bitfire.icsdroid.MiscUtils;
import at.bitfire.icsdroid.R;
import lombok.Cleanup;

public class AddCalendarValidationFragment extends DialogFragment implements LoaderManager.LoaderCallbacks<ResourceInfo> {
    private static final String ARG_INFO = "info";

    public static AddCalendarValidationFragment newInstance(@NonNull ResourceInfo info) {
        AddCalendarValidationFragment frag = new AddCalendarValidationFragment();
        Bundle args = new Bundle(1);
        args.putSerializable(ARG_INFO, info);
        frag.setArguments(args);
        return frag;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        getLoaderManager().initLoader(0, getArguments(), this);
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        setCancelable(false);

        ProgressDialog progress = new ProgressDialog(getActivity());
        progress.setMessage(getString(R.string.add_calendar_validating));
        return progress;
    }


    // loader callbacks

    @Override
    public Loader<ResourceInfo> onCreateLoader(int id, Bundle args) {
        return new ResourceInfoLoader(getContext(), (ResourceInfo)args.getSerializable(ARG_INFO));
    }

    @Override
    public void onLoadFinished(Loader<ResourceInfo> loader, ResourceInfo info) {
        getDialog().dismiss();

        String errorMessage = null;
        if (info.exception != null)
            errorMessage = info.exception.getMessage();
        else if (info.statusCode != HttpURLConnection.HTTP_OK)
            errorMessage = info.statusCode + " " + info.statusMessage;

        if (errorMessage == null) {
            if (StringUtils.isEmpty(info.calendarName))
                info.calendarName = info.url.getFile();

            getFragmentManager()
                    .beginTransaction()
                    .replace(R.id.fragment_container, AddCalendarDetailsFragment.newInstance(info))
                    .addToBackStack(null)
                    .commitAllowingStateLoss();
        } else
            Toast.makeText(getActivity(), errorMessage, Toast.LENGTH_SHORT).show();
    }

    @Override
    public void onLoaderReset(Loader<ResourceInfo> loader) {
    }


    // loader

    protected static class ResourceInfoLoader extends AsyncTaskLoader<ResourceInfo> {
        ResourceInfo info;
        boolean loaded;

        public ResourceInfoLoader(Context context, ResourceInfo info) {
            super(context);
            this.info = info;
        }

        @Override
        protected void onStartLoading() {
            synchronized(this) {
                if (!loaded) {
                    forceLoad();
                    loaded = true;
                }
            }
        }

        @Override
        public ResourceInfo loadInBackground() {
            info.exception = null;

            URLConnection conn = null;

            URL url = info.url;
            boolean followRedirect = true;
            for (int redirect = 0; followRedirect && redirect < Constants.MAX_REDIRECTS; redirect++) {
                followRedirect = false;
                try {
                    conn = url.openConnection();
                    conn.setConnectTimeout(7000);
                    conn.setReadTimeout(20000);

                    if (conn instanceof HttpsURLConnection)
                        MTMLoader.prepareHttpsURLConnection(getContext(), (HttpsURLConnection)conn);

                    if (conn instanceof HttpURLConnection) {
                        HttpURLConnection httpConn = (HttpURLConnection)conn;
                        conn.setRequestProperty("User-Agent", Constants.USER_AGENT);
                        httpConn.setInstanceFollowRedirects(false);

                        if (info.authRequired) {
                            String basicCredentials = info.username + ":" + info.password;
                            conn.setRequestProperty("Authorization", "Basic " + Base64.encodeToString(basicCredentials.getBytes(), 0));

                        }
                        info.statusCode = httpConn.getResponseCode();
                        info.statusMessage = httpConn.getResponseMessage();

                        // handle redirects
                        String location = httpConn.getHeaderField("Location");
                        if (info.statusCode/100 == 3 && location != null) {
                            Log.i(Constants.TAG, "Following redirect to " + location);
                            url = new URL(url, location);
                            followRedirect = true;
                            if (info.statusCode == HttpURLConnection.HTTP_MOVED_PERM) {
                                Log.i(Constants.TAG, "Permanent redirect: saving new location");
                                info.url = url;
                            }
                        }

                        // only read stream if status is 200 OK
                        if (info.statusCode != HttpURLConnection.HTTP_OK)
                            conn = null;

                    } else
                        // local file, always simulate HTTP status 200 OK
                        info.statusCode = HttpURLConnection.HTTP_OK;

                } catch (IOException e) {
                    info.exception = e;
                }
            }

            try {
                if (conn != null) {
                    @Cleanup InputStreamReader reader = new InputStreamReader(conn.getInputStream(),
                            MiscUtils.charsetFromContentType(conn.getContentType()));
                    Map<String, String> properties = new HashMap<>();
                    List<Event> events = Event.fromReader(reader, properties);

                    info.calendarName = properties.get(Event.CALENDAR_NAME);
                    info.eventsFound = events.size();
                }

            } catch (IOException|InvalidCalendarException e) {
                info.exception = e;
            }
            return info;
        }
    }
}

File: app/src/main/java/at/bitfire/icsdroid/ui/AddCalendarValidationFragment.kt
/***************************************************************************************************
 * Copyright © All Contributors. See LICENSE and AUTHORS in the root directory for details.
 **************************************************************************************************/

package at.bitfire.icsdroid.ui

import android.app.Application
import android.app.Dialog
import android.app.ProgressDialog
import android.net.Uri
import android.os.Bundle
import android.util.Log
import androidx.core.content.ContextCompat
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import androidx.fragment.app.viewModels
import androidx.lifecycle.*
import at.bitfire.ical4android.Css3Color
import at.bitfire.ical4android.Event
import at.bitfire.ical4android.ICalendar
import at.bitfire.icsdroid.CalendarFetcher
import at.bitfire.icsdroid.Constants
import at.bitfire.icsdroid.HttpClient
import at.bitfire.icsdroid.HttpUtils.toURI
import at.bitfire.icsdroid.HttpUtils.toUri
import at.bitfire.icsdroid.R
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import net.fortuna.ical4j.model.property.Color
import okhttp3.MediaType
import java.io.InputStream
import java.io.InputStreamReader

@Deprecated("Use Jetpack Compose")
class AddCalendarValidationFragment: DialogFragment() {

    private val titleColorModel by activityViewModels<TitleColorFragment.TitleColorModel>()
    private val credentialsModel by activityViewModels<CredentialsFragment.CredentialsModel>()

    private val validationModel by viewModels<ValidationModel> {
        object : ViewModelProvider.Factory {
            @Suppress("UNCHECKED_CAST")
            override fun <T : ViewModel> create(modelClass: Class<T>): T {
                val uri = Uri.parse(titleColorModel.url.value ?: throw IllegalArgumentException("No URL given"))!!
                val authenticate = credentialsModel.requiresAuth.value ?: false
                return ValidationModel(
                    requireActivity().application,
                    uri,
                    if (authenticate) credentialsModel.username.value else null,
                    if (authenticate) credentialsModel.password.value else null
                ) as T
            }
        }
    }


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        validationModel.result.observe(this, Observer { info ->
            requireDialog().dismiss()

            val exception = info.exception
            if (exception == null) {
                titleColorModel.url.value = info.uri.toString()

                if (titleColorModel.color.value == null)
                    titleColorModel.color.value = info.calendarColor ?: ContextCompat.getColor(requireContext(), R.color.lightblue)

                if (titleColorModel.title.value.isNullOrBlank())
                    titleColorModel.title.value = info.calendarName ?: info.uri.toString()

                parentFragmentManager
                    .beginTransaction()
                    .replace(android.R.id.content, AddCalendarDetailsFragment())
                    .addToBackStack(null)
                    .commitAllowingStateLoss()
            } else {
                val errorMessage =
                    exception.localizedMessage ?: exception.message ?: exception.toString()
                AlertFragment.create(errorMessage, exception).show(parentFragmentManager, null)
            }
        })
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val progress = ProgressDialog(activity)
        progress.setMessage(getString(R.string.add_calendar_validating))
        return progress
    }

    override fun onPause() {
        super.onPause()
        HttpClient.setForeground(false)
    }

    override fun onResume() {
        super.onResume()
        HttpClient.setForeground(true)
    }


    /* activityModel and data source */

    class ValidationModel(
            val context: Application,
            val originalUri: Uri,
            val username: String?,
            val password: String?
    ): ViewModel() {

        val result = MutableLiveData<ResourceInfo>()

        init {
            viewModelScope.launch(Dispatchers.Default) {
                validate()
            }
        }

        private suspend fun validate() {
            Log.i(Constants.TAG, "Validating Webcal feed $originalUri (authentication: $username)")

            val info = ResourceInfo(originalUri)
            val downloader = object: CalendarFetcher(context, originalUri) {
                override suspend fun onSuccess(data: InputStream, contentType: MediaType?, eTag: String?, lastModified: Long?, displayName: String?) {
                    InputStreamReader(data, contentType?.charset() ?: Charsets.UTF_8).use { reader ->
                        val properties = mutableMapOf<String, String>()
                        val events = Event.eventsFromReader(reader, properties)

                        info.calendarName = properties[ICalendar.CALENDAR_NAME] ?: displayName
                        info.calendarColor =
                            // try COLOR first
                            properties[Color.PROPERTY_NAME]?.let { colorValue ->
                                Css3Color.colorFromString(colorValue)
                            } ?:
                            // try X-APPLE-CALENDAR-COLOR second
                            try {
                                properties[ICalendar.CALENDAR_COLOR]?.let { colorValue ->
                                    Css3Color.colorFromString(colorValue)
                                }
                            } catch (e: IllegalArgumentException) {
                                Log.w(Constants.TAG, "Couldn't parse calendar COLOR", e)
                                null
                            }
                        info.eventsFound = events.size
                    }

                    result.postValue(info)
                }

                override suspend fun onNewPermanentUrl(target: Uri) {
                    Log.i(Constants.TAG, "Got permanent redirect when validating, saving new URL: $target")
                    val location = uri.toURI().resolve(target.toURI())
                    info.uri = location.toUri()
                }

                override suspend fun onError(error: Exception) {
                    Log.e(Constants.TAG, "Couldn't validate calendar", error)
                    info.exception = error
                    result.postValue(info)
                }
            }

            downloader.username = username
            downloader.password = password

            // directly ask for confirmation of custom certificates
            downloader.inForeground = true

            downloader.fetch()
        }

    }

}